File: packages/perspective/src/js/api/view_api.js
/******************************************************************************
 *
 * Copyright (c) 2019, the Perspective Authors.
 *
 * This file is part of the Perspective library, distributed under the terms of
 * the Apache License 2.0.  The full license can be found in the LICENSE file.
 *
 */

import {subscribe, unsubscribe, async_queue} from "./dispatch.js";

/**
 * Construct a proxy for the view object by creating a "view" message and
 * sending it through the worker.
 *
 * @param {*} worker
 * @param {*} table_name
 * @param {*} config
 */
export function view(worker, table_name, config) {
    return new Promise((resolve, reject) => {
        this._worker = worker;
        this._name = Math.random() + "";

        this._worker.post(
            {
                cmd: "view",
                view_name: this._name,
                table_name: table_name,
                config: config
            },
            () => {
                // Resolve the Promise with this function, which is a proxy
                // view that dispatches all view methods to the backend, i.e.
                // a Web Worker or a Python/Node Perspective server. Because
                // arrow functions automatically capture `this` from the
                // surrounding scope, there is no need to explicitly bind the
                // resolving value to `this`.
                resolve(this);
            },
            reject
        );
    });
}

/**
 * Create a reference to a view located on `worker` for use by remote clients.
 *
 * @param {worker} worker the Web Worker at which the view is located.
 * @param {String} name a unique name for the view.
 */
export function proxy_view(worker, name) {
    this._worker = worker;
    this._name = name;
}

proxy_view.prototype = view.prototype;

// Send view methods that do not create new objects (getters, setters etc.) to
// the queue for processing.

view.prototype.get_config = async_queue("get_config");

view.prototype.to_json = async_queue("to_json");

view.prototype.to_arrow = async_queue("to_arrow");

view.prototype.to_columns = async_queue("to_columns");

view.prototype.to_csv = async_queue("to_csv");

view.prototype.schema = async_queue("schema");

view.prototype.computed_schema = async_queue("computed_schema");

view.prototype.column_paths = async_queue("column_paths");

view.prototype.num_columns = async_queue("num_columns");

view.prototype.num_rows = async_queue("num_rows");

view.prototype.set_depth = async_queue("set_depth");

view.prototype.get_row_expanded = async_queue("get_row_expanded");

view.prototype.expand = async_queue("expand");

view.prototype.collapse = async_queue("collapse");

view.prototype.delete = async_queue("delete");

view.prototype.col_to_js_typed_array = async_queue("col_to_js_typed_array");

view.prototype.on_update = subscribe("on_update", "view_method", true);

view.prototype.remove_update = unsubscribe("remove_update", "view_method", true);

view.prototype.on_delete = subscribe("on_delete", "view_method", true);

view.prototype.remove_delete = unsubscribe("remove_delete", "view_method", true);

File: python/perspective/perspective/client/view_api.py
#  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
#  ┃ ██████ ██████ ██████       █      █      █      █      █ █▄  ▀███ █       ┃
#  ┃ ▄▄▄▄▄█ █▄▄▄▄▄ ▄▄▄▄▄█  ▀▀▀▀▀█▀▀▀▀▀ █ ▀▀▀▀▀█ ████████▌▐███ ███▄  ▀█ █ ▀▀▀▀▀ ┃
#  ┃ █▀▀▀▀▀ █▀▀▀▀▀ █▀██▀▀ ▄▄▄▄▄ █ ▄▄▄▄▄█ ▄▄▄▄▄█ ████████▌▐███ █████▄   █ ▄▄▄▄▄ ┃
#  ┃ █      ██████ █  ▀█▄       █ ██████      █      ███▌▐███ ███████▄ █       ┃
#  ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
#  ┃ Copyright (c) 2017, the Perspective Authors.                              ┃
#  ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
#  ┃ This file is part of the Perspective library, distributed under the terms ┃
#  ┃ of the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0). ┃
#  ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

import asyncio

from random import random
from functools import partial
from .dispatch import async_queue, subscribe, unsubscribe


def view(
    client,
    table_name,
    columns=None,
    group_by=None,
    split_by=None,
    aggregates=None,
    sort=None,
    filter=None,
    expressions=None,
):
    """Create a new View by posting a message to the Perspective server
    implementation through `client`, returning a Future that will resolve to a
    `PerspectiveViewProxy` object whose API must be called with `await` or
    `yield`, or an Exception if the View creation failed.
    """
    name = str(random())

    config = {
        "columns": columns,
        "group_by": group_by,
        "split_by": split_by,
        "aggregates": aggregates,
        "sort": sort,
        "filter": filter,
        "expressions": expressions,
    }

    msg = {
        "cmd": "view",
        "view_name": name,
        "table_name": table_name,
        "config": config,
    }

    future = asyncio.Future()
    client.post(msg, future)
    return future


class PerspectiveViewProxy(object):
    def __init__(self, client, name):
        """A proxy for a Perspective `View` object elsewhere, i.e. on a remote
        server accessible through a Websocket.

        All public API methods on this proxy are async, and must be called
        with `await` or a `yield`-based coroutine.

        Args:
            client (:obj:`PerspectiveClient`): A `PerspectiveClient` that is
                set up to send messages to a Perspective server implementation
                elsewhere.

            name (:obj:`str`): a `str` name for the View. Automatically
                generated if using the `view` function defined above.
        """
        self._client = client
        self._name = name
        self._async_queue = partial(async_queue, self._client, self._name)
        self._subscribe = partial(subscribe, self._client, self._name)
        self._unsubscribe = partial(unsubscribe, self._client, self._name)

    def get_config(self):
        return self._async_queue("get_config", "view_method")

    def sides(self):
        return self._async_queue("sides", "view_method")

    def num_rows(self):
        return self._async_queue("num_rows", "view_method")

    def num_columns(self):
        return self._async_queue("num_columns", "view_method")

    def get_min_max(self):
        return self._async_queue("get_min_max", "view_method")

    def get_row_expanded(self, idx):
        return self._async_queue("get_row_expanded", "view_method", idx)

    def expand(self, idx):
        return self._async_queue("expand", "view_method", idx)

    def collapse(self, idx):
        return self._async_queue("collapse", "view_method", idx)

    def set_depth(self, idx):
        return self._async_queue("set_depth", "view_method", idx)

    def column_paths(self):
        return self._async_queue("column_paths", "view_method")

    def schema(self, as_string=False):
        return self._async_queue("schema", "view_method", as_string=as_string)

    def expression_schema(self, as_string=False):
        return self._async_queue("expression_schema", "view_method", as_string=as_string)

    def on_update(self, callback, mode=None):
        return self._subscribe("on_update", "view_method", callback, mode=mode)

    def remove_update(self, callback):
        return self._unsubscribe("remove_update", "view_method", callback)

    def on_delete(self, callback):
        return self._subscribe("on_delete", "view_method", callback)

    def remove_delete(self, callback):
        return self._unsubscribe("remove_delete", "view_method", callback)

    def delete(self):
        return self._async_queue("delete", "view_method")

    def to_arrow(self, **kwargs):
        return self._async_queue("to_arrow", "view_method", **kwargs)

    def to_records(self, **kwargs):
        return self._async_queue("to_records", "view_method", **kwargs)

    def to_dict(self, **kwargs):
        return self._async_queue("to_dict", "view_method", **kwargs)

    def to_numpy(self, **kwargs):
        return self._async_queue("to_numpy", "view_method", **kwargs)

    def to_df(self, **kwargs):
        return self._async_queue("to_df", "view_method", **kwargs)

    def to_csv(self, **kwargs):
        return self._async_queue("to_csv", "view_method", **kwargs)

    def to_json(self, **kwargs):
        return self._async_queue("to_json", "view_method", **kwargs)

    def to_columns(self, **kwargs):
        return self._async_queue("to_columns", "view_method", **kwargs)

    def to_columns_string(self, **kwargs):
        return self._async_queue("to_columns_string", "view_method", **kwargs)
